### ISSUE TITLE
Incorrect error when reassigning to narrowed generic variable

### ISSUE THREAD

**Issue by sisp** [python/mypy#13868]
**Bug Report**

I've encountered an error when reassigning to a narrowed generic variable. See the following [reproducible example](https://mypy-play.net/?mypy=latest&python=3.10&gist=bff5ef8ada5cdf5ae3212a1c5b0d2777):

```python
from typing import TypeVar


T = TypeVar("T")


def f(value: T) -> T:
    if isinstance(value, str):
        # This shouldn't throw a mypy error because `value` is a `builtins.str`
        # here.
        value = value + ""

    return value
```

**To Reproduce**

```python
# Ideally, a small sample program that demonstrates the problem.
# Or even better, a reproducible playground link https://mypy-play.net/ (use the "Gist" button)
```

**Expected Behavior**

The assignment of a string to `value` should be fine because in that if-condition `value` has been narrowed to a `builtins.str`, also according to mypy.

**Actual Behavior**

```
main.py:11: error: Incompatible types in assignment (expression has type "str", variable has type "T")
```

**Your Environment**

See the [mypy-play.net example](https://mypy-play.net/?mypy=latest&python=3.10&gist=bff5ef8ada5cdf5ae3212a1c5b0d2777).

**Comment by AlexWaygood** [comment]
Indeed:

```pycon
>>> from typing import TypeVar
>>> T = TypeVar("T")
>>> class Custom(str):
...     def __add__(self, other: str) -> int:
...         return 42
...
>>> def f(value: T) -> T:
...     if isinstance(value, str):
...         value = value + ""
...     return value
...
>>> x = Custom("foo")
>>> print(type(x))
<class '__main__.Custom'>
>>> type(f(x))
<class 'int'>
```

**Comment by AlexWaygood** [comment]
Actually -- @erictraut, would it not be okay to solve the type variable as `str` in this case, given that `str` would be a common base class of `str` and `CustomStr`?

**Comment by sisp** [comment]
@erictraut Ah, you're right! :+1: Thanks for the clarification. :pray: I hadn't thought about the subclass case.

**Comment by sisp** [comment]
Just wondering, what would be the correct way of implementing this example? Using `if type(value) is str: ...` instead of `if isinstance(value, str): ...`?

```python
>>> from typing import TypeVar
>>> T = TypeVar("T")
>>> class CustomStr(str): ...
...
>>> def f(value: T) -> T:
...     if type(value) is str: # <------ like this?
...         value = value + ""
...     return value
...
>>> x = CustomStr("foo")
>>> type(x)
<class '__main__.CustomStr'>
>>> type(f(x))
<class '__main__.CustomStr'>
```
